File: 01_USWTDB_to_PIDs.py
''' ============================
Import packages and set directory
============================ '''
import os, sys
import pandas as pd
import geopandas as gpd
from shapely.geometry import box
from shapely.geometry import Polygon
from shapely.geometry import Point

thisDir = os.path.abspath(os.curdir)
print('Working directory:  ', thisDir)

''' ============================
Read and crop USWTDB point geometries
============================ '''

wind_proj = pd.read_csv(os.path.join(thisDir, 'data', 'uswtdb', 'uswtdb_v4_3_20220114_ucsb.csv'))

wind_proj_geom = gpd.GeoDataFrame(wind_proj, 
                                  geometry = gpd.points_from_xy(wind_proj['xlong'], wind_proj['ylat']),
                                  crs = 'EPSG:4269')

# Create bounding box for contiguous U.S.
bbox = box(-125, 24, -66.5, 50)

# Crop geometries for the contiguous US
wind_proj_contiguous = gpd.clip(wind_proj_geom, mask = bbox)

# Write GeoDataFrame to a shapefile
# output_shapefile_path = os.path.join(thisDir, 'data', 'existing_wind_contiguous_us.shp')
# wind_proj_contiguous.to_file(output_shapefile_path)

''' ============================
Find USWTDB project centroids
============================ '''

# Group by p_id and keep p_cap for SAM simulation and optimization
wind_proj = wind_proj_contiguous[['p_id', 'p_cap', 'geometry']]
wind_proj = wind_proj.dissolve(by = 'p_id', aggfunc = 'first').reset_index()

# Rename column
wind_proj = wind_proj.rename(columns = {'0': 'geometry'})  

# Create polygons from point geometries and find wind project centroids
centroid = wind_proj.dissolve(by = 'p_id').convex_hull.centroid.reset_index().dropna() 

# Join centroids to variable of interest (p_cap)
p_cap = wind_proj[['p_id', 'p_cap']]
pid_site_coords = pd.merge(p_cap, centroid, on = 'p_id')

# Rename column
pid_site_coords = pid_site_coords.rename(columns = {'0': 'centroid'})  

# Reset index
pid_coords = pid_site_coords.reset_index()

# Rename column
pid_coords.columns.values[3] = 'geometry'

# Set type as GeoDataFrame
pid_coords_geom = gpd.GeoDataFrame(pid_coords,geometry = pid_coords.columns[3], crs = 'EPSG:4269')

# Extract longtiude from centroid
pid_coords_geom['lon'] = pid_coords_geom.geometry.x

# Extract latitude from centroid
pid_coords_geom['lat'] = pid_coords_geom.geometry.y

# Subset and save df to csv
pid_coords_geom = pid_coords_geom[['index', 'p_cap', 'lon', 'lat']].rename(columns = {'index': 'PID'})
file_path = os.path.join(thisDir, 'data', 'uswtdb', 'us_PID_cords.csv')
pid_coords.to_csv(file_path, index = False)

''' ============================
Associate NREL GEA region to PID
============================ '''

File: 01_USWTDB_to_PIDs.R
library(here)
library(tidyverse)
library(sf)
library(tmap)
library(tmaptools)

# Define filepath
uswtdb_dir <- here::here('data', 'uswtdb')

# Read USWTDB
wind_proj <- read_csv(paste0(uswtdb_dir, '/uswtdb_v4_3_20220114_ucsb.csv'))

# Add spatial coordinates
wind_proj_geom <- sf::st_as_sf(wind_proj, coords = c('xlong', 'ylat'), crs = 4269)

# class(wind_proj_geom$geometry)

# Create bounding box for contiguous U.S.
contiguous_us <- sf::st_bbox(c(xmin = -125, ymin = 24, xmax = -66.5, ymax = 50), crs = st_crs(wind_proj_geom))
wind_proj_contiguous <- sf::st_crop(wind_proj_geom, contiguous_us)

# View 
# tm_shape(wind_proj_contiguous) +
#  tm_dots(col = '#2A9D8F') +
#  tm_layout(main.title = 'Exisiting Wind Sites in the Contiguous U.S.')

# Save as shapefile
# st_write(wind_proj_contiguous, here::here('data', 'existing_wind_contiguous_us.shp'))

# Group by project ID 
pid_data <- wind_proj_contiguous %>% 
  group_by(p_id) %>%
  # Keep project capacity
  summarise(p_cap = first(p_cap)) %>%
  # Create polygon from point data
  mutate(polygon = st_convex_hull(geometry)) %>%
  # Find centroid of polygon
  mutate(centroid = st_centroid(polygon)) %>%
  filter(!is.na(p_id))

# Set view to interactive
# tmap_mode(mode = 'view')

# Plot to check that centroids/polygons are accurate
# tm_shape(pid_data$geometry) +
#  tm_dots(col = '#264653', size = 0.2) +
#  tm_shape(pid_data$polygon) +
#  tm_fill(col = '#2a9d8f', alpha = 0.5) +
#  tm_shape(pid_data$centroid) +
#  tm_dots(col = '#e76f51')

# Select variables needed for SAM simulation
pid_site_coords <- pid_data %>%
  select(c(p_cap, centroid)) %>% 
  mutate(centroid = st_transform(centroid, crs = 4269))

# Subset and drop spatial class
pid_coords <- pid_site_coords %>%
  mutate(PID = row_number()) %>%
  mutate(lon = unlist(map(pid_site_coords$centroid, 1)),
         lat = unlist(map(pid_site_coords$centroid, 2))) %>% 
  select(c(PID, p_cap, lon, lat)) %>% 
  sf::st_drop_geometry()

# Save df to csv
write_csv(pid_coords, paste0(uswtdb_dir, '/us_PID_cords.csv'), row.names = FALSE)